Dialog to choose from multiple template files per task in wrap

kalisp opened the pull request:
This PR adds possibility to assign multiple Wrap templates for single task, as it is recommended to split Wrap logic into more workfiles for technical reasons.

This functionality is configured in `ayon+settings://wrap/multiple_templates_per_tasks?project=ayon_test` where multiple `steps` could be configured for profile of task name and type.

![image](https://github.com/user-attachments/assets/41a6ca7f-9bb0-4502-9b5e-40081e049683)

If matching profile is found during start of Wrap, artist is shown simple Workfile app that shows
- list of configured templates in left column
- existing workfiles created from this template
App allows creation (copy of template to workarea) of new workfile, or selection of any existing workfile.
![image](https://github.com/user-attachments/assets/9a5a4e49-1107-43a3-b172-d961e946fe8b)

This functionality is not required, if single workfile template per task is enough, ordinary `ayon+settings://wrap/workfile_builder?project=ayon_test` should be used.

kalisp committed:
Updated obsolete Field to SettingsField

kalisp committed:
Added Settings for multiple templates per task

kalisp committed:
Fix name of field

kalisp committed:
Added hook to trigger Wrap Workfile tool

Checks if current context has configuration for multiple templates. If not it skips Workfiles tool.

kalisp committed:
Yanked logic to lib

Will be used in Workfiles tool to get templates.
I prefer simpler interface for the tool, profile could be passed directly too, but it seems nicer to run the tool only on the context and let prepare data itself.

kalisp committed:
Added workfile path template for multiple workfiles

Added here as it is depending of multi profiles, `ayon+anatomy://ayon_test/templates/work` would be used even if not multi templates should be used.
`ayon+settings://core/tools/Workfiles/workfile_template_profiles?project=ayon_test` would duplicate tasks that should have multi templates.

It is not expected that there would be need for multiple path templates

kalisp committed:
Added model and widget for workarea files

Tried to inherit from WorkAreaFilesModel from Workfile app, didn't work. Had to copy and cleaned up code from there.

kalisp committed:
Added window with templates and workarea widgets

kalisp committed:
Added FileItem for workfiles

kalisp committed:
Basic controller to fill templates and workarea widgets

kalisp committed:
Added creation of first workfile

kalisp committed:
Added passing of workfile path up

Workfile app sets env var to propagate new or opening existing workfile.

kalisp committed:
Removed comment from workfile

<_> pattern is problematic for basic formatting and there is no actual field to pass comment in.

kalisp committed:
Added closing app when Open or Create New

kalisp committed:
Preselect first template

kalisp committed:
Cleanup

kalisp committed:
Set state of buttons

Checks presence of workfiles in workarea to set enable state for Open and Create New buttons

kalisp committed:
Set height of buttons

kalisp committed:
Close app after create

kalisp committed:
Fix proper usage of workfile in arguments

launch_args contains old path to workfile, not from Workfile app for multiple templates per task.

kalisp committed:
Change ApplicationLaunchFailed to PlaceholderFillException

Error in filling of placeholder shouldn't stop launching of application.

iLLiCiTiT commented on client/ayon_wrap/hooks/pre_multiple_templates.py:
```diff
@@ -0,0 +1,62 @@
+                "task_id": task_entity["id"]
+            }
+            workfiles_tool = WorkfilesToolWindow(launch_data=launch_data)
+            workfiles_tool.show()
+            workfiles_tool.raise_()
+            workfiles_tool.activateWindow()
+            workfiles_tool.showNormal()
+
```
You don't need to call any of them if you call `exec_()`
```suggestion
```

iLLiCiTiT commented on client/ayon_wrap/hooks/pre_multiple_templates.py:
```diff
@@ -0,0 +1,62 @@
+            workfiles_tool.exec_()
+            workfile_path = os.environ.get("WRAP_WORKFILE_PATH")
+            if not workfile_path or not os.path.exists(workfile_path):
+                raise RuntimeError(f"'{workfile_path} doesn't exist!")
```
```suggestion
                raise ApplicationLaunchFailed(f"'{workfile_path} doesn't exist!")
```

iLLiCiTiT committed:
moved window to widgets

iLLiCiTiT committed:
refactored widgets to use controller as controller and removed unnecessary widgets

iLLiCiTiT committed:
modified controller to do what controller should do

kalisp committed:
Add WorkfilesToolWindow to be directly imported

kalisp committed:
Fix retrieval of existing template item

iLLiCiTiT committed:
use wrapper widget

iLLiCiTiT committed:
fix items loop

iLLiCiTiT committed:
Merge branch 'develop' into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

iLLiCiTiT committed:
Merge branch 'develop' into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

# Conflicts:
#	client/ayon_wrap/api/lib.py
#	client/ayon_wrap/hooks/pre_replace_placeholders.py

iLLiCiTiT committed:
fix settings linting

iLLiCiTiT committed:
remove unused import

kalisp committed:
Reverted from Wrap settings to Anatomy template

After review comment, it should be in Anatomy, added to documentation and switched to pulling and formatting it from Anatomy.

kalisp committed:
Reverted adding WorkfilesToolWindow to init

This way it should be safer against circular imports

kalisp committed:
Decreased size of dialog

kalisp commented:
Added documentation PR https://github.com/ynput/ayon-documentation/pull/202

kalisp committed:
Merge branch 'feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap' of https://github.com/ynput/ayon-wrap into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

iLLiCiTiT commented on client/ayon_wrap/workfiles/control.py:
```diff
@@ -37,6 +37,9 @@ def __init__(self, launch_data):
         project_entity = ayon_api.get_project(project_name)
 
+        # template name in Anatomy
+        self.wrap_template_name = "wrap_multi"
```
If you've already change it to use anatomy then use `get_workfile_template_key` to determine which should be used.

kalisp replied:
This workflow is applicable only for specific profiles, to be able to use Anatomy, I would need to duplicate that logic there.

iLLiCiTiT replied:
I don't understand, you are using anatomy, but also hardcoding template name... You should not hardcode template name, you should use whatever is defined in workfile settings, and studio should add the template (they should be able to name it whatever they want) and add it to profiles for `wrap` host.

iLLiCiTiT commented on client/ayon_wrap/workfiles/widgets/window.py:
```diff
@@ -0,0 +1,209 @@
+        template_names = self._controller.get_template_names()
+        templates_widget = self._templates_widget
+        existing_items = {}
+        for idx in templates_widget.count():
+            item = templates_widget.item(idx)
+            if not item:
+                break
```
```suggestion
        for idx in range(templates_widget.count()):
            item = templates_widget.item(idx)
```

iLLiCiTiT committed:
fox items collection

kalisp committed:
Fix iterating through items

kalisp committed:
Fix button state

Button state must be refreshed after each template_change.
Proxy model counted it wrong (empty items included).

kalisp committed:
Merge branch 'feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap' of https://github.com/ynput/ayon-wrap into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

iLLiCiTiT committed:
don't use private attributes

iLLiCiTiT committed:
don't use controller reset event

iLLiCiTiT committed:
simplified template name changed topic

iLLiCiTiT committed:
removed unused variable

kalisp committed:
Update exception

Co-authored-by: Jakub Trllo <43494761+iLLiCiTiT@users.noreply.github.com>

kalisp committed:
Remove unnecessary functions

exec_() should be enough

Co-authored-by: Jakub Trllo <43494761+iLLiCiTiT@users.noreply.github.com>

kalisp committed:
Merge branch 'develop' into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

kalisp committed:
Add comments why traypublisher there

No real Wrap host, creator and publish can only run in Traypublisher.

kalisp committed:
Merge branch 'feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap' of https://github.com/ynput/ayon-wrap into feature/AY-5403_Provide-dialog-to-choose-from-multiple-template-files-per-task-in-Wrap

kalisp committed:
Switch from hardcoding template key to profiles

kalisp committed:
Changed label

kalisp committed:
Changed label
